core/win32/drwinapi/ntdll_redir.c

```diff
@@ -833,6 +833,23 @@ redirect_RtlPcToFileHeader(__in PVOID PcValue, __out PVOID *BaseOfImage)
     LOG(GLOBAL, LOG_LOADER, 2, "%s " PFX " => " PFX "\n", __FUNCTION__, PcValue, res);
     return res;
 }
+/* XXX i#3633: Isolate PEB->Fls* for Windows 10 1903 where Fls is stored inside of
+ * ntdll.dll internal variables. This code isolates PEB->Fls* in dynamorio global
+ * variable, thus we don't use PEB anymore for FLS handling
+ */
+#ifndef FLS_MAX_COUNT
+#    define FLS_MAX_COUNT 128
+#endif
+
+typedef struct _FLS_GLOB_DATA {
+    RTL_BITMAP *FlsBitmap;
+    LIST_ENTRY FlsListHead;
+    PVOID *FlsCallback;
+    ULONG FlsHighIndex;
+    ULONG FlsBitmapBits[5];
```

Comment:
Our ntdll.h lists this as `DWORD FlsBitmapBits[4]`.  Why 5 instead of 4 here?  Is this extended to 5 in 1903?  Worth a comment.

Reply:
Correct, it should be 4. I've made type error. However, Windows 10 1903 apparently allows up to 0xFF0 FLS slots, according to this code from ntdll.dll:

```
.text:4B305F63                 cmp     eax, 0FF0h
.text:4B305F68                 jnb     loc_4B305FEF
```

compared to the old code 

```
.text:0DCE7CF8                 cmp     eax, 7Eh ; '~'
.text:0DCE7CFB                 ja      loc_DD0B617
```